telemetry/Tracks/class-tracks.php

```diff
@@ -0,0 +1,79 @@
+<?php
+/**
+ * Telemetry: Tracks class
+ *
+ * @package Automattic\VIP\Telemetry
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry;
+
+use WP_Error;
+
+/**
+ * This class comprises the mechanics of sending events to the Automattic Tracks
+ * system.
+ */
+class Tracks extends Telemetry_System {
+	/**
+	 * Registers the events into WordPress hooks to activate tracking.
+	 */
+	public function run(): void {
+		$this->activate_tracking();
+	}
+
+	/**
+	 * Records an event to Tracks by using the Tracks pixel.
+	 *
+	 * Depending on the current context, the pixel will be recorded
+	 * synchronously (as a GET request) or as asynchronously (as an injected
+	 * pixel into the page's footer).
+	 *
+	 * If the event doesn't pass validation, it gets silently discarded.
+	 *
+	 * @param string                            $event_name The event name. Must be snake_case.
+	 * @param array<string, mixed>|array<empty> $event_properties Any additional properties to include with the event.
+	 *                                                            Key names must be lowercase and snake_case.
+	 * @return bool|WP_Error True if recording the event succeeded.
+	 *                       False if telemetry is disabled.
+	 *                       WP_Error if recording the event failed.
+	 */
+	public function record_event(
+		string $event_name,
+		array $event_properties = array()
+	) {
+		$event = new Tracks_Event( $event_name, $event_properties );
+		$pixel = Tracks_Client::instance();
+
+		// Process AJAX/REST request events immediately.
+		if ( wp_doing_ajax() || defined( 'REST_REQUEST' ) ) {
+			$pixel->record_event_synchronously( $event );
+		}
+
+		return $pixel->record_event_asynchronously( $event );
+	}
+
+	/**
+	 * Registers the events into their respective WordPress hooks, so they
+	 * can be recorded when the hook fires.
+	 */
+	protected function activate_tracking(): void {
```

Comment:
I think we should get rid of this abstraction and make all our our tracking explicit via `record_event()` calls.

Reply:
We would still need a way register all the callbacks to the appropriate filters and `record_event()` is specifically used by the callbacks to send the events to the Tracks system so I'm not sure to achieve what you're asking 🤔

Reply:
What I mean is that rather than abstracting like so: https://github.com/Automattic/vip-go-mu-plugins/blob/44e6293555331be77646e30ee2df97c961fcd34c/vip-parsely/vip-parsely.php#L37-L55

The way it works now, the registration of the hook is separated too far away the actual tracking code which makes it harder to debug.

We should let plugins do the extra work needed to record what they need:

```
add_action( 'transition_post_status', function( $new, $old, $post ) {
    Automattic\VIP\Telemetry::get_instance()->record_event( 'post_status_changed', [ ... ] );
}, 10, 3 );
```

This requires more code but keeps the code more explicit and easier to read / follow / debug.

Reply:
@Automattic/vip-platform-cantina what do y'all think?

Reply:
I agree with @mjangda , and I don't even think this will be more code? 

And if a particular plugin wants to enrich the data somehow it can do something like where it would make sure that the data has the necessary event properies

```php
Automattic\VIP\Telemetry::get_instance()->record_event( 'post_status_changed', enrich_track_event( [ ... ] ) );
```

Reply:
Agreed on the above.

This also leads to a good point - we should add a README for this library with an example of how to use it. That'll help to simplify it down once you see how a plugin would implement it. IMO, it should be as easy to follow as the `send_pixel` method.

Reply:
Removed in 637891038f65b5a66d5bff9840f68d36e6d1ffd3